### ISSUE TITLE
Preview Feedback

### ISSUE THREAD

**Issue by daveverwer** [SwiftPackageIndex/SwiftPackageIndex-Server#300]
Thank you for being part of our secret (🤫) little preview for the Swift Package Index. 🚀

(Yes yes, I know this is all 100% available to the public, and if anyone looked at mine or @finestructure's GitHub profile pages this would be the only significant activity. Luckily no one cares!)

Please leave any feedback you feel happy to make in public here on this thread so everyone can see it. If you'd like to email me private feedback, you can do so at dave@verwerservices.com.

While you might suggest something we already have in our backlog, don't feel you need to read through everything there before commenting here. Your reasons for wanting stuff are valuable, even if we already have the feature logged. Thank you so much in advance to for your feedback.

The site is launching next Friday, just as a sense of the timescale.

Thanks
Dave & Sven

**Comment by daveverwer** [comment]
What feedback are we seeking? Of course, any feedback at all is welcome, but if you want some guidance:

* Does the site work, make sense, and provide useful information about available Swift packages?
* Does the site help you to make decisions about which packages might be worth your time to investigate further?

**Comment by MaxDesiatov** [comment]
Thanks for your work, this looks great! One thing I noticed is the "Language and Platforms" section for each package. For example, [for CoreXLSX](https://swiftpackageindex.com/CoreOffice/CoreXLSX) it says

> ## Language and Platforms
>    Versions 0.11.0 and master support:
>      iOS 9.0+, macOS 10.11+, watchOS 2.0+, and tvOS 9.0+.

This is for a library that supports Linux, as I test multiple Swift versions on Linux on CI. In my opinion, the way the section currently looks is somewhat misleading. For this package, it seems like the library doesn't support Linux at all.

As far as I understand, you're parsing the `platforms` array in `Package.swift`. The name for that array argument is misleading too, it should have been called `platformRequirements`, as this doesn't indicate platforms that are support, but only requirements for platforms if any requirements are present. Based on this, I think, the section should at least be renamed to something like "Platform Requirements". It's also not clear what "Language" bit in the current section name means.

**Comment by MaxDesiatov** [comment]
In terms of decision making, it would be wonderful if the number of direct and the total number of transitive dependencies for a package were displayed. Additionally, I hope the package page would show the number of contributors, not just the number of commits. It also seems that the package licence is not displayed anywhere, which also has a direct impact on whether people take the package as a new dependency.

**Comment by daveverwer** [comment]
> This is for a library that supports Linux, as I test multiple Swift versions on Linux on CI. In my opinion, the way the section currently looks is somewhat misleading. For this package, it seems like the library doesn't support Linux at all.

I agree completely, and this was a problem with the SwiftPM Library implementation too. As we can't (easily) add metadata to the Package.swift file, what we're planning is to propose a new metadata file. Initially unofficial, but if it gains traction, it may become official by ubiquity. Ideas for it are being tracked in #202 and you'll see that a Linux support boolean is the first thing on the list.

We're planning to put a proposal for the metdata that file might include out to the community before implementing it, but I can say with certainty that metadata around Linux support will make it in.

> As far as I understand, you're parsing the `platforms` array in `Package.swift`. The name for that array argument is misleading too, it should have been called `platformRequirements`, as this doesn't indicate platforms that are supported, but only requirements for platforms if any requirements are present. Based on this, I think the section should at least be renamed to something like "Platform Requirements".

This is an interesting point, but my gut feeling says that "support" is probably slightly less misleading than "requirements". The [documentation](https://developer.apple.com/documentation/swift_packages/package/3197886-platforms) specifically calls out that you're declaring "support" for the platforms too.

Ultimately, those flags are entirely at the discretion of the package author, so it's how they choose to fill them in rather than being anything formal. I think once people are able to also specify Linux as a supported platform (from above) this issue will be much less of a problem.

**Comment by MaxDesiatov** [comment]
I'm not so sure about Linux being a boolean. Now with support for multiple Linux distributions, you probably would want to make this metadata richer, e.g. I only test my packages on Ubuntu 18.04, but I don't think that they work on Ubuntu 14.04 and I didn't test it anything on Ubuntu 16.04, or 20.04, or Amazon Linux 2, which Swift now also supports. 

Additionally, the number of platforms will grow quickly, Swift 5.3 is supposed to support Windows, there are plenty of packages for Android, and I personally going to added WebAssembly support for my packages and would love the catalog to indicate that. In fact, there's already a precedent for [a package that only supports WebAssembly](https://github.com/kateinoigakukun/JavaScriptKit/).

**Comment by mattt** [comment]
I'm overjoyed to see this latest evolution of your work to provide the Swift community a much-needed index of available packages. A hearty congratulations to you, @daveverwer, and to @finestructure on this launch.

I hope to share some more structured feedback in coming days, but here are some quick observations:

- I like that search is prioritized, shown right at the top. But I think the landing page would also benefit from showing more packages all at once.
- "Recent Packages" should probably be called "Recently Added" or something else ("packages" seems redundant here)
- Under "Recent Releases", each entry should say which version was released.
- The representation for an unknown license (such as on [this page](https://swiftpackageindex.com/3lvis/Networking)) uses a color of green that indicates success, when perhaps a muted gray color might be more appropriate to indicate indeterminate status. It also incorporates the [OSI logo](https://opensource.org/node/442) which may be misleading if the unknown license is indeed non-compliant with OSI's [definition of "open source"](https://opensource.org/osd).

<img width="233" alt="Screen Shot 2020-06-06 at 12 29 21" src="https://user-images.githubusercontent.com/7659/83952951-7c255600-a7f1-11ea-9456-5f2d1d147687.png">

- I think it'd be helpful to create a taxonomy to tag / categorize the functionality of packages (e.g. "Networking", "UI", "Database"). No small task, but I see some opportunities for automating a lot of the process using text classifiers or analyzing API usage.
- There's a small pluralization issue with "There are 1 open issue" on [this page](https://swiftpackageindex.com/vapor-community/HTMLKit)
- The `_blank` target on the link to "Submit a Package" was surprising. I think that should open in the same tab / window. You may also consider using an icon to indicate that this is an external link (which is unexpected for something in the header navigation).

**Comment by MaxDesiatov** [comment]
@daveverwer

>  The documentation specifically calls out that you're declaring "support" for the platforms too.

I'd consider that documentation misleading too (and we have plenty of precedents of Apple docs being incomplete or incorrect), [here's what Apple folks say directly](https://forums.swift.org/t/platforms-stanza-for-linux/34955/9) in an answer to a suggestion to make `platforms` an explicit list of supported platforms:

>we’ll be changing the interpretation of a pre-existing stanza across a tools version: the platforms directive used to have an implicit “accept all others”, now it has an implicit “reject all others”. This is going to surprise some folks!

> Finally, note that changing to implicit reject makes adopting new platforms harder. For example, it means the entire Swift ecosystem now claims it doesn’t work on Windows, even though large swathes of it do. That’s a difficult case to justify: how do users opt out of this?

The whole thread is worth reading, the gist is that it would make sense to rename `platforms` to `platformRequirements`, but it's not done to avoid compatibility issues:
> My worry would be that renaming the API will make the situation even more confusing, because packages using a tools-version prior to the rename will of course continue to use the old name. It seems like that would reinforce the idea that platforms has a meaning that it does not.

**Comment by daveverwer** [comment]
> It's also not clear what exactly the "Language" bit in the current section name means, I haven't seen any information related to it. If you displayed the minimum Swift version in `Package.swift` that would make more sense. If you do try to determine the minimum Swift version, please don't forget to parse additional manifests like `Package@swift-4.2.swift` that some libraries provide to maintain backward compatibility.

There's currently a bug with the site where if you supply platform information, but do not specify the `swiftLanguageVersions` key, it looks like this and doesn't even mention the Swift language version.

If you have *no* language/platform metadata then it [looks like this](https://swiftpackageindex.com/JohnSundell/Files) and points you to [this FAQ article](https://swiftpackageindex.com/faq#language-and-platforms). I've just realised that even though I'm aware of that bug, I don't think it's logged anywhere so I just logged it as #301

The SwiftPM Library did much more parsing of Swift language information and tried not only to grab it out of the manifest, but also looked for Xcode projects and started inspecting them! We decided with the SPI to keep it simpler and rely on the manifest only. This is definitely more limited and starkly shows how few packages declare the `swiftLanguageVersions ` property. We may enhance this later, initially through #191, but maybe further after that.

**Comment by MaxDesiatov** [comment]
I personally don't think there's much incentive to specify `swiftLanguageVersions` in `Package.swift` at all. If my manifest specifies 5.0, I imply that it supports 5.0 and later. And to provide support for 4.2 I have a separate `Package@swift-4.2.swift` manifest, especially because both manifests use different SwiftPM features (`platforms` argument is not available in 4.2).

Anyway, I don't want to digress to SwiftPM discussions, sorry for polluting the thread with technicalities. If there's a separate issue to discuss package manifests parsing, I'd be happy to chime in though.

**Comment by daveverwer** [comment]
> I'm not so sure about Linux being a boolean. Now with support for multiple Linux distributions, you probably would want to make this metadata richer, e.g. I only test my packages on Ubuntu 18.04, but I don't think that they work on Ubuntu 14.04 and I didn't test it anything on Ubuntu 16.04, or 20.04, or Amazon Linux 2, which Swift now also supports.

This is a really good point, and one that I think is worth discussing in much more detail on #202. @MaxDesiatov, would you mind chiming in over there. We haven't discussed what data format this file would be yet (there are a few choices) and that decision will partly be informed by the data we need to track. I'd love to get your input on what data you'd like to see associated with my vague "Linux support" aspect of that metadata. Thanks!

**Comment by daveverwer** [comment]
> @daveverwer
> 
> > The documentation specifically calls out that you're declaring "support" for the platforms too.
> 
> I'd consider that documentation misleading too (and we have plenty of precedents of Apple docs being incomplete or incorrect), [here's what Apple folks say directly](https://forums.swift.org/t/platforms-stanza-for-linux/34955/9) in an answer to a suggestion to make `platforms` an explicit list of supported platforms:
> 
> > we’ll be changing the interpretation of a pre-existing stanza across a tools version: the platforms directive used to have an implicit “accept all others”, now it has an implicit “reject all others”. This is going to surprise some folks!
> 
> > Finally, note that changing to implicit reject makes adopting new platforms harder. For example, it means the entire Swift ecosystem now claims it doesn’t work on Windows, even though large swathes of it do. That’s a difficult case to justify: how do users opt out of this?
> 
> The whole thread is worth reading, the gist is that it would make sense to rename `platforms` to `platformRequirements`, but it's not done to avoid compatibility issues:
> 
> > My worry would be that renaming the API will make the situation even more confusing, because packages using a tools-version prior to the rename will of course continue to use the old name. It seems like that would reinforce the idea that platforms has a meaning that it does not.
> 
> People in that same thread also clarify why specifying "support Linux" doesn't make sense, because of how Linux distributions are different, even among versions of the same distribution.

I just read that thread, and the thread linked in that one too and it kinda made me want to just dodge the whole issue. 😬 This really needs some serious thought. Thank you so much for your feedback on this @MaxDesiatov.

**Comment by daveverwer** [comment]
> I'm overjoyed to see this latest evolution of your work to provide the Swift community a much-needed index of available packages. A hearty congratulations to you, @daveverwer, and to @finestructure on this launch.

Thank you @mattt!

> * I like that search is prioritized, shown right at the top. But I think the landing page would also benefit from showing more packages all at once.

Yep, I can see straight away some kind of “Top Packages” driven by stars. The only disadvantage of this list is that is rarely changes. Both Recently added and Recent releases move much more quickly. Are there any specific things you’d want to see here? Also, having a set of 4 would be nicer than 3, for layout reasons :) 

> * "Recent Packages" should probably be called "Recently Added" or something else ("packages" seems redundant here)

This is much better. Thank you. Fixed in #304

> * Under "Recent Releases", each entry should say which version was released.

100% This is in progress as part of the work for #64

> * The representation for an unknown license (such as on [this page](https://swiftpackageindex.com/3lvis/Networking)) uses a color of green that indicates success, when perhaps a muted gray color might be more appropriate to indicate indeterminate status. It also incorporates the [OSI logo](https://opensource.org/node/442) which may be misleading if the unknown license is indeed non-compliant with OSI's [definition of "open source"](https://opensource.org/osd).

The colour is in progress as #204 and should be done before Friday’s launch.

What I hadn’t considered though is that icon! I’ll also make sure to remove that in the specific case of an Unknown License. I also want to add a FAQ entry on how to fix that being “Unknown license” to point it to GitHub’s documentation on it. That’s being tracked as #203.

> * I think it'd be helpful to create a taxonomy to tag / categorize the functionality of packages (e.g. "Networking", "UI", "Database"). No small task, but I see some opportunities for automating a lot of the process using text classifiers or analyzing API usage.

I agree 100%, and there are a few options here. #1 is to let package authors define categories in a metadata file (#202). #2 is to lean into GitHub’s tags through their API, many packages have tagged themselves in this way. #3 is auto-analysis, which is great because it’s automated, but less great because it’s automated 😂

In all likelihood, it’s a combination of all three that will yield the best results. I created an Epic for it, so we can track this as an overall “thing” to implement - #305

> * There's a small pluralization issue with "There are 1 open issue" on [this page](https://swiftpackageindex.com/vapor-community/HTMLKit)

Well spotted! Thank you #302

> * The `_blank` target on the link to "Submit a Package" was surprising. I think that should open in the same tab/window. You may also consider using an icon to indicate that this is an external link (which is unexpected for something in the header navigation).

Yep, there’s a blanket rule to open all external links in a new tab (which I like) but this shouldn’t really be an external link. Added as #306.

Thanks again, and please do let us know if you have more feedback!

**Comment by daveverwer** [comment]
> In terms of decision making, it would be wonderful if the number of direct and the total number of transitive dependencies for a package were displayed. Additionally, I hope the package page would show the number of contributors, not just the number of commits. It also seems that the package licence is not displayed anywhere, which has a direct impact on whether people take the package as a new dependency.
> 
> For direct and transitive dependencies you'd probably need to do some smart parsing of targets/products arrays, so that dependencies of test targets are not included in the dependencies count.

Dependency exploration and their display is something @finestructure have talked about several times, but it's never made it as far as a logged issue. Now it has its own epic as #307

**Comment by gspiers** [comment]
## General

I think some of my comments below might have been addressed already. 

Looking great. The site is so fast and responsive. There's also a lot of information presented in a simple way.

When looking for a package to solve a problem I can see myself using this site first to create a short list list of packages to investigate in depth. The stats are a really easy way to quickly determine if a package is even worth investigating. Further investigation I tend to do is read some of the source code, and look for tests.

I can’t remember if you index packages with Licenses that are not safe to use in closed source apps? If so, I’d like to be able to narrow my search to exclude these.

Other stats I’d like to see (not sure you can easily get these):
* Which Apple frameworks does the package link to. I like to avoid third party dependencies that use UIKit as an example.
* Does the package contain any swizzling code.
* How many dependencies does the package itself have. Would be great to be able to have another section of the package’s dependencies that link to the package detail page on your site.
* Is there a test target.

## Desktop

* Recent section looks a bit sparse, can the description or version go here (like the search result)?
* Nice privacy policy (is it worth pointing out you aren’t legally compliant?)
* The ‘Add a Package’ might be better linking to the `#adding-packages` anchor on the Readme on Github.
* FAQ “Packages are also polled for changes every couple of hours, so the information you see in the index always knows about the latest package versions, without the package author needing to tell the index about every new release.” sentence is a bit long, also sounds a bit weird that the ‘index knows’. Maybe “Packages are also polled for changes every couple of hours, so the information you see in the index always reflects the latest package versions all without any action from the package author.”
* FAQ: Words run together “There’s a proposalfor”
* FAQ: “The Swift Package Index will support and index any GitHub’s, or any other implementation of a package registry as soon as they are available.” Reads funny “any GitHub’s” Maybe “The Swift Package Index will support and index any GitHub’s Package Registry, and any other implementation of a package registry as soon as they are available.”
* I was going to [tease you](https://twitter.com/daveverwer/status/1042710234651938816) that each package detail page needs a how do I add the URL to my `Package.swift` file 😉. But honestly, a link to a tutorial on the FAQ might be helpful for people. Probably need a “How do I use these packages?” FAQ (someone will ask).
* Package detail page: Nice touch would be a copy URL button, or even a copy button that gives me the exact string to paste into my `Package.swift`. Example for Alamofire the clipboard would contain: `.package(url: "https://github.com/Alamofire/Alamofire.git", .upToNextMajor(from: “5.2.1”))`

## Mobile

* Mobile: Search dropdown viewport is wider and allows scrolling horizontally. Same issue on FAQ on my iPhone XS.
* Mobile: Home screen looks sparse below the search.

**Comment by daveverwer** [comment]
Thank you @gspiers !

> and look for tests.

This is on the list, but wasn’t yet an issue. Filed as #317.

> I can’t remember if you index packages with Licenses that are not safe to use in closed source apps? If so, I’d like to be able to narrow my search to exclude these.

This was on the list, but not written down anywhere. Filed as #318.

> * Which Apple frameworks does the package link to. I like to avoid third party dependencies that use UIKit as an example.

This is great feedback, and I can see it being something for the future but it feels a bit of a ways out from where we are right now with the analysis. I’m going to keep a note of it, but not add it to the backlog for now.

> * Does the package contain any swizzling code.

Same

> * How many dependencies does the package itself have. Would be great to be able to have another section of the package’s dependencies that link to the package detail page on your site.

Absolutely. #307

> * Recent section looks a bit sparse, can the description or version go here (like the search result)?

I looked at this, but some descriptions are really long! I could clip it, but ended up deciding it was better without. The metadata file (#202) may give us some better options here.

> * Nice privacy policy (is it worth pointing out you aren’t legally compliant?)

Fixed in #316

> * The ‘Add a Package’ might be better linking to the `#adding-packages` anchor on the Readme on Github.

Even better, this is going to be a page on the site itself #306

> * FAQ “Packages are also polled for changes every couple of hours, so the information you see in the index always knows about the latest package versions, without the package author needing to tell the index about every new release.” sentence is a bit long, also sounds a bit weird that the ‘index knows’. Maybe “Packages are also polled for changes every couple of hours, so the information you see in the index always reflects the latest package versions all without any action from the package author.”

Yea this was nonsense! Looks like I blacked out half way through writing that sentence and then picked up elsewhere when I woke up.

Fixed in #321

> * FAQ: Words run together “There’s a proposalfor”

The best kind of feedback is feedback that comes with a pull request! 🎉 Thanks for the fix!

> * FAQ: “The Swift Package Index will support and index any GitHub’s, or any other implementation of a package registry as soon as they are available.” Reads funny “any GitHub’s” Maybe “The Swift Package Index will support and index any GitHub’s Package Registry, and any other implementation of a package registry as soon as they are available.”

Fixed in #321

> * I was going to [tease you](https://twitter.com/daveverwer/status/1042710234651938816) that each package detail page needs a how do I add the URL to my `Package.swift` file 😉. But honestly, a link to a tutorial on the FAQ might be helpful for people. Probably need a “How do I use these packages?” FAQ (someone will ask).

Fixed in #321

> * Package detail page: Nice touch would be a copy URL button, or even a copy button that gives me the exact string to paste into my `Package.swift`. Example for Alamofire the clipboard would contain: `.package(url: "https://github.com/Alamofire/Alamofire.git", .upToNextMajor(from: “5.2.1”))`

Yep, this is on the list. Sticking with just a copy button for now. It’s logged as #74.

> * Mobile: Search dropdown viewport is wider and allows scrolling horizontally. Same issue on FAQ on my iPhone XS.

I can’t reproduce this, time to dig out my iPhone X! Logged as #320

**Comment by chriseidhof** [comment]
This looks great! Some very quick not well-thought out feedback:

I'd like more ways to quickly judge whether a package is good. Clearly, there's nothing better than understanding the source code / trying it out, but there are some indicators: I know that certain authors write good packages, and if a package has a lot of stars on Github, it's also a clear indicator that at least a lot of people are looking at it. It could be interesting to include these two things (they are just the first two that come to mind).

Another interesting thing I'd like to see is what the library depends on. For me personally, if a library depends on too many other things, I don't want to use it. (Typically this is an indication that the library is fragile).

There are just my opinions, I completely acknowledge that other people might have very different viewpoints so don't optimize for me. If anything, I've come to use way *less* third-party code over the years, so I might not be the target audience anwyay.

Thanks for making this 💖

**Comment by daveverwer** [comment]
> This looks great! Some very quick not well-thought out feedback:

Thank you Chris!

> I know that certain authors write good packages

Absolutely. Author support is absolutely on the way (#173, #174, and #175)

> If a package has a lot of stars on Github, it's also a clear indicator that at least a lot of people are looking at it. It could be interesting to include these two things (they are just the first two that come to mind).

How on earth did I forget to include this in the designs?!? We even grab that number of stars already, and include it as part of our scoring mechanism for sort order. How did it not make it through to the home page?! Logged as #338.

> Another interesting thing I'd like to see is what the library depends on. For me personally, if a library depends on too many other things, I don't want to use it. (Typically this is an indication that the library is fragile).

Yep, absolutely. This is in the backlog as #307 

> There are just my opinions, I completely acknowledge that other people might have very different viewpoints so don't optimize for me. If anything, I've come to use way _less_ third-party code over the years, so I might not be the target audience anwyay.

No, this is all stuff that I agree 100% with! Using potentially less, but higher quality third-party code is what this site is all about.

> Thanks for making this 💖

Thanks for your feedback!

**Comment by daveverwer** [comment]
Thank you all for your fantastic feedback here, and privately. The launch was a great success today, and your feedback helped make that the case!

Thank you 😍
